Multiline property wrappers cause false positives for `let_var_whitespace`

### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Given the following code, `let_var_whitespace` will fire at lines 4 and 8.

```
internal struct Foo {
    @Flag(name: [.long, .customLong("autocorrect")], help: "Correct violations whenever possible.")
    internal var fix = false
    @Flag(help: """
        Should reformat the Swift files using the same mechanism used by Xcode (via SourceKit).
        Only applied with `--fix`/`--autocorrect`.
        """)
    internal var format = false
    @Flag(help: "Use an alternative algorithm to exclude paths for `excluded`, which may be faster in some cases.")
    internal var useAlternativeExcluding = false
}
```

This code is actually from LintOrAnalyzeArguments.swift

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
% swiftlint lint --no-cache --enable-all-rules Foo.swift
Linting Swift files at paths Foo.swift
Linting 'Foo.swift' (1/1)
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
/tmp/Foo.swift:3:14: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
/tmp/Foo.swift:8:14: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
/tmp/Foo.swift:10:14: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
/tmp/Foo.swift:4:5: warning: Variable Declaration Whitespace Violation: Let and var should be separated from other statements by a blank line. (let_var_whitespace)
/tmp/Foo.swift:8:5: warning: Variable Declaration Whitespace Violation: Let and var should be separated from other statements by a blank line. (let_var_whitespace)
Done linting! Found 5 violations, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file:

none

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

no

* Which Xcode version are you using (check `xcodebuild -version`)?

```
Xcode 14.2
Build version 14C18
```

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

See above.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,11 @@
 * Rewrite `multiline_literal_brackets` rule using SwiftSyntax, fixing some
   false positives that would happen when comments are present.  
   [Marcelo Fabri](https://github.com/marcelofabri)
+  
+* Rewrite `let_var_whitespace` rule using SwiftSyntax, fixing false positives
+  when attributes attached to declarations were spread over multiple lines.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#4801](https://github.com/realm/SwiftLint/pull/4801)
 
 #### Bug Fixes
 

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/DeploymentTargetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/DeploymentTargetRule.swift
@@ -3,6 +3,7 @@ import SwiftSyntax
 @SwiftSyntaxRule
 struct DeploymentTargetRule {
     fileprivate typealias Version = DeploymentTargetConfiguration.Version
+
     var configuration = DeploymentTargetConfiguration()
 
     static let description = RuleDescription(

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift
@@ -1,8 +1,5 @@
 import SwiftLintCore
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
-
 @AutoApply
 struct DiscouragedDirectInitConfiguration: SeverityBasedRuleConfiguration {
     typealias Parent = DiscouragedDirectInitRule

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/IndentationWidthConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/IndentationWidthConfiguration.swift
@@ -1,8 +1,5 @@
 import SwiftLintCore
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
-
 @AutoApply
 struct IndentationWidthConfiguration: SeverityBasedRuleConfiguration {
     typealias Parent = IndentationWidthRule

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/TestCaseAccessibilityConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/TestCaseAccessibilityConfiguration.swift
@@ -1,8 +1,5 @@
 import SwiftLintCore
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
-
 @AutoApply
 struct TestCaseAccessibilityConfiguration: SeverityBasedRuleConfiguration {
     typealias Parent = TestCaseAccessibilityRule

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/UnitTestConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/UnitTestConfiguration.swift
@@ -5,9 +5,6 @@ typealias EmptyXCTestMethodConfiguration = UnitTestConfiguration<EmptyXCTestMeth
 typealias SingleTestClassConfiguration = UnitTestConfiguration<SingleTestClassRule>
 typealias NoMagicNumbersConfiguration = UnitTestConfiguration<NoMagicNumbersRule>
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
-
 @AutoApply
 struct UnitTestConfiguration<Parent: Rule>: SeverityBasedRuleConfiguration {
     @ConfigurationElement(key: "severity")

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/UnusedDeclarationConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/UnusedDeclarationConfiguration.swift
@@ -1,8 +1,5 @@
 import SwiftLintCore
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
-
 @AutoApply
 struct UnusedDeclarationConfiguration: SeverityBasedRuleConfiguration {
     typealias Parent = UnusedDeclarationRule

--- a/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
@@ -138,6 +138,23 @@ struct LetVarWhitespaceRule: OptInRule {
                 #if os(macOS)
                 let a = 2
                 #endif
+            """),
+            // Don't trigger in closure bodies.
+            Example("""
+                f {
+                    let a = 1
+                    return a
+                }
+            """),
+            Example("""
+                func f() {
+                    #if os(macOS)
+                    let a = 2
+                    return a
+                    #else
+                    return 1
+                    #endif
+                }
             """)
         ],
         triggeringExamples: [
@@ -177,7 +194,16 @@ struct LetVarWhitespaceRule: OptInRule {
             Example("""
                 let a = 2
                 ↓b = 1
-            """)
+            """),
+            Example("""
+                #if os(macOS)
+                let a = 0
+                ↓func f() {}
+                #else
+                func f() {}
+                ↓let a = 1
+                #endif
+            """, excludeFromDocumentation: true)
         ]
     )
 
@@ -193,7 +219,7 @@ private extension LetVarWhitespaceRule {
         }
 
         override func visitPost(_ node: CodeBlockItemListSyntax) {
-            if node.parent?.is(SourceFileSyntax.self) == true {
+            if node.isInValidContext {
                 collectViolations(from: node, using: { $0.unwrap })
             }
         }
@@ -218,6 +244,22 @@ private extension LetVarWhitespaceRule {
                 }
             }
         }
+    }
+}
+
+private extension CodeBlockItemListSyntax {
+    var isInValidContext: Bool {
+        var next = parent
+        while let ancestor = next {
+            if [.closureExpr, .codeBlock, .accessorBlock].contains(ancestor.kind) {
+                return false
+            }
+            if [.memberBlock, .sourceFile].contains(ancestor.kind) {
+                return true
+            }
+            next = ancestor.parent
+        }
+        return false
     }
 }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
@@ -10,25 +10,85 @@ struct LetVarWhitespaceRule: OptInRule {
         description: "Variable declarations should be separated from other statements by a blank line",
         kind: .style,
         nonTriggeringExamples: [
-            Example("let a = 0\nvar x = 1\n\nx = 2"),
-            Example("a = 5\n\nvar x = 1"),
-            Example("struct X {\n\tvar a = 0\n}"),
-            Example("let a = 1 +\n\t2\nlet b = 5"),
-            Example("var x: Int {\n\treturn 0\n}"),
-            Example("var x: Int {\n\tlet a = 0\n\n\treturn a\n}"),
-            Example("#if os(macOS)\nlet a = 0\n#endif"),
-            Example("#warning(\"TODO: remove it\")\nlet a = 0"),
-            Example("#error(\"TODO: remove it\")\nlet a = 0"),
-            Example("@available(swift 4)\nlet a = 0"),
-            Example("class C {\n\t@objc\n\tvar s: String = \"\"\n}"),
-            Example("class C {\n\t@objc\n\tfunc a() {}\n}"),
-            Example("class C {\n\tvar x = 0\n\tlazy\n\tvar y = 0\n}"),
-            Example("@available(OSX, introduced: 10.6)\n@available(*, deprecated)\nvar x = 0"),
             Example("""
-            // swiftlint:disable superfluous_disable_command
-            // swiftlint:disable force_cast
+                let a = 0
+                var x = 1
 
-            let x = bar as! Bar
+                x = 2
+            """),
+            Example("""
+                a = 5
+
+                var x = 1
+            """),
+            Example("""
+                struct X {
+                    var a = 0
+                }
+            """),
+            Example("""
+                let a = 1 +
+                    2
+                let b = 5
+            """),
+            Example("""
+                var x: Int {
+                    return 0
+                }
+            """),
+            Example("""
+                var x: Int {
+                    let a = 0
+
+                    return a
+                }
+            """),
+            Example("""
+                #if os(macOS)
+                let a = 0
+                #endif
+            """),
+            Example("""
+                #warning("TODO: remove it")
+                let a = 0
+            """),
+            Example("""
+                #error("TODO: remove it")
+                let a = 0
+            """),
+            Example("""
+                @available(swift 4)
+                let a = 0
+            """),
+            Example("""
+                class C {
+                    @objc
+                    var s: String = ""
+                }
+            """),
+            Example("""
+                class C {
+                    @objc
+                    func a() {}
+                }
+            """),
+            Example("""
+                class C {
+                    var x = 0
+                    lazy
+                    var y = 0
+                }
+            """),
+            Example("""
+                @available(OSX, introduced: 10.6)
+                @available(*, deprecated)
+                var x = 0
+            """),
+            Example("""
+                // swiftlint:disable superfluous_disable_command
+                // swiftlint:disable force_cast
+
+                let x = bar as! Bar
             """),
             Example("""
                 @available(swift 4)
@@ -38,25 +98,54 @@ struct LetVarWhitespaceRule: OptInRule {
                 @Attribute
                 func f() {}
             """),
-            Example("var x: Int {\n\tlet a = 0\n\treturn a\n}"), // don't trigger on local vars
+            // Don't trigger on local variable declarations.
             Example("""
-            struct S {
-                static var test: String { /* Comment block */
-                    let s = "!"
-                    return "Test" + s
+                var x: Int {
+                    let a = 0
+                    return a
                 }
+            """),
+            Example("""
+                struct S {
+                    static var test: String { /* Comment block */
+                        let s = "!"
+                        return "Test" + s
+                    }
 
-                func f() {}
-            }
+                    func f() {}
+                }
             """, excludeFromDocumentation: true)
         ],
         triggeringExamples: [
-            Example("var x = 1\n↓x = 2"),
-            Example("\na = 5\n↓var x = 1"),
-            Example("struct X {\n\tlet a\n\t↓func x() {}\n}"),
-            Example("var x = 0\n↓@objc func f() {}"),
-            Example("var x = 0\n↓@objc\n\tfunc f() {}"),
-            Example("@objc func f() {\n}\n↓var x = 0"),
+            Example("""
+                var x = 1
+                ↓x = 2
+            """),
+            Example("""
+
+                a = 5
+                ↓var x = 1
+            """),
+            Example("""
+                struct X {
+                    let a
+                    ↓func x() {}
+                }
+            """),
+            Example("""
+                var x = 0
+                ↓@objc func f() {}
+            """),
+            Example("""
+                var x = 0
+                ↓@objc
+                func f() {}
+            """),
+            Example("""
+                @objc func f() {
+                }
+                ↓var x = 0
+            """),
             Example("""
                 struct S {
                     func f() {}

--- a/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
@@ -1,6 +1,7 @@
-import Foundation
-import SourceKittenFramework
+import SwiftLintCore
+import SwiftSyntax
 
+@SwiftSyntaxRule
 struct LetVarWhitespaceRule: OptInRule {
     var configuration = SeverityConfiguration<Self>(.warning)
 
@@ -14,17 +15,15 @@ struct LetVarWhitespaceRule: OptInRule {
                 let a = 0
                 var x = 1
 
-                x = 2
+                var y = 2
             """),
             Example("""
-                a = 5
+                let a = 5
 
                 var x = 1
             """),
             Example("""
-                struct X {
-                    var a = 0
-                }
+                var a = 0
             """),
             Example("""
                 let a = 1 +
@@ -46,11 +45,14 @@ struct LetVarWhitespaceRule: OptInRule {
             Example("""
                 #if os(macOS)
                 let a = 0
+                func f() {}
                 #endif
             """),
             Example("""
                 #warning("TODO: remove it")
                 let a = 0
+                #warning("TODO: remove it")
+                let b = 0
             """),
             Example("""
                 #error("TODO: remove it")
@@ -61,23 +63,17 @@ struct LetVarWhitespaceRule: OptInRule {
                 let a = 0
             """),
             Example("""
-                class C {
-                    @objc
-                    var s: String = ""
-                }
+                @objc
+                var s: String = ""
             """),
             Example("""
-                class C {
-                    @objc
-                    func a() {}
-                }
+                @objc
+                func a() {}
             """),
             Example("""
-                class C {
-                    var x = 0
-                    lazy
-                    var y = 0
-                }
+                var x = 0
+                lazy
+                var y = 0
             """),
             Example("""
                 @available(OSX, introduced: 10.6)
@@ -106,31 +102,29 @@ struct LetVarWhitespaceRule: OptInRule {
                 }
             """),
             Example("""
-                struct S {
-                    static var test: String { /* Comment block */
-                        let s = "!"
-                        return "Test" + s
-                    }
-
-                    func f() {}
+                static var test: String { /* Comment block */
+                    let s = "!"
+                    return "Test" + s
                 }
-            """, excludeFromDocumentation: true)
-        ],
+
+                func f() {}
+            """, excludeFromDocumentation: true),
+            Example(#"""
+                @Flag(name: "name", help: "help")
+                var fix = false
+                @Flag(help: """
+                        help
+                        text
+                """)
+                var format = false
+                @Flag(help: "help")
+                var useAlternativeExcluding = false
+            """#, excludeFromDocumentation: true)
+        ].map(Self.wrapIntoClass),
         triggeringExamples: [
             Example("""
-                var x = 1
-                ↓x = 2
-            """),
-            Example("""
-
-                a = 5
-                ↓var x = 1
-            """),
-            Example("""
-                struct X {
-                    let a
-                    ↓func x() {}
-                }
+                let a
+                ↓func x() {}
             """),
             Example("""
                 var x = 0
@@ -147,228 +141,52 @@ struct LetVarWhitespaceRule: OptInRule {
                 ↓var x = 0
             """),
             Example("""
-                struct S {
-                    func f() {}
-                    ↓@Wapper
-                    let isNumber = false
-                    @Wapper
-                    var isEnabled = true
-                    ↓func g() {}
-                }
+                func f() {}
+                ↓@Wapper
+                let isNumber = false
+                @Wapper
+                var isEnabled = true
+                ↓func g() {}
             """)
-        ]
+        ].map(Self.wrapIntoClass)
     )
 
-    func validate(file: SwiftLintFile) -> [StyleViolation] {
-        let dict = file.structureDictionary
-
-        var attributeLines = attributeLineNumbers(file: file)
-        var varLines = Set<Int>()
-        varLetLineNumbers(file: file,
-                          structure: dict.substructure,
-                          attributeLines: &attributeLines,
-                          collectingInto: &varLines)
-        let skippedLines = skippedLineNumbers(file: file)
-        var violations = [StyleViolation]()
-
-        for (index, line) in file.lines.enumerated() {
-            guard !varLines.contains(index) &&
-                  !skippedLines.contains(index) else {
-                continue
-            }
-
-            let trimmed = line.content.trimmingCharacters(in: .whitespaces)
-            guard trimmed.isNotEmpty else {
-                continue
-            }
-
-            // Precedes var/let and has text not ending with {
-            if linePrecedesVar(index, varLines, skippedLines) {
-                if !trimmed.hasSuffix("{") &&
-                   !file.lines[index + 1].content.trimmingCharacters(in: .whitespaces).hasPrefix("}") {
-                    violated(&violations, file, index + 1)
-                }
-            }
-            // Follows var/let and has text not starting with }
-            if lineFollowsVar(index, varLines, skippedLines) {
-                if !trimmed.hasPrefix("}") &&
-                   !file.lines[index - 1].content.trimmingCharacters(in: .whitespaces).hasSuffix("{") {
-                    violated(&violations, file, index)
-                }
-            }
-        }
-        return violations
-    }
-
-    private func linePrecedesVar(_ lineNumber: Int, _ varLines: Set<Int>, _ skippedLines: Set<Int>) -> Bool {
-        return lineNeighborsVar(lineNumber, varLines, skippedLines, 1)
-    }
-
-    private func lineFollowsVar(_ lineNumber: Int, _ varLines: Set<Int>, _ skippedLines: Set<Int>) -> Bool {
-        return lineNeighborsVar(lineNumber, varLines, skippedLines, -1)
-    }
-
-    private func lineNeighborsVar(_ lineNumber: Int, _ varLines: Set<Int>,
-                                  _ skippedLines: Set<Int>, _ increment: Int) -> Bool {
-        if varLines.contains(lineNumber + increment) {
-            return true
-        }
-
-        var prevLine = lineNumber
-
-        while skippedLines.contains(prevLine) {
-            if varLines.contains(prevLine + increment) {
-                return true
-            }
-            prevLine += increment
-        }
-        return false
-    }
-
-    private func violated(_ violations: inout [StyleViolation], _ file: SwiftLintFile, _ line: Int) {
-        let content = file.lines[line].content
-        let startIndex = content.rangeOfCharacter(from: CharacterSet.whitespaces.inverted)?.lowerBound
-                         ?? content.startIndex
-        let offset = content.distance(from: content.startIndex, to: startIndex)
-        let location = Location(file: file, characterOffset: offset + file.lines[line].range.location)
-
-        violations.append(StyleViolation(ruleDescription: Self.description,
-                                         severity: configuration.severity,
-                                         location: location))
-    }
-
-    private func lineOffsets(file: SwiftLintFile, statement: SourceKittenDictionary) -> (Int, Int)? {
-        guard let offset = statement.offset,
-              let length = statement.length
-        else {
-            return nil
-        }
-        let startLine = file.line(byteOffset: offset)
-        let endLine = file.line(byteOffset: offset + length)
-
-        return (startLine, endLine)
-    }
-
-    // Collects all the line numbers containing var or let declarations
-    private func varLetLineNumbers(file: SwiftLintFile,
-                                   structure: [SourceKittenDictionary],
-                                   attributeLines: inout Set<Int>,
-                                   collectingInto result: inout Set<Int>) {
-        for statement in structure {
-            guard statement.kind != nil,
-                  let (startLine, endLine) = lineOffsets(file: file, statement: statement) else {
-                continue
-            }
-
-            if let declarationKind = statement.declarationKind {
-                if SwiftDeclarationKind.nonVarAttributableKinds.contains(declarationKind) {
-                    if attributeLines.contains(startLine) {
-                        attributeLines.remove(startLine)
-                    }
-                }
-                if SwiftDeclarationKind.varKinds.contains(declarationKind) {
-                    var lines = Set(startLine...((endLine < 0) ? file.lines.count : endLine))
-                    var previousLine = startLine - 1
-
-                    // Include preceding attributes
-                    while attributeLines.contains(previousLine) {
-                        lines.insert(previousLine)
-                        attributeLines.remove(previousLine)
-                        previousLine -= 1
-                    }
-
-                    // Exclude the body where the accessors are
-                    if let bodyOffset = statement.bodyOffset,
-                        let bodyLength = statement.bodyLength {
-                        let bodyStart = file.line(byteOffset: bodyOffset) + 1
-                        let bodyEnd = file.line(byteOffset: bodyOffset + bodyLength) - 1
-
-                        if bodyStart <= bodyEnd {
-                            lines.subtract(Set(bodyStart...bodyEnd))
-                        }
-                    }
-                    result.formUnion(lines)
-                }
-            }
-
-            let substructure = statement.substructure
-
-            if substructure.isNotEmpty {
-                varLetLineNumbers(file: file,
-                                  structure: substructure,
-                                  attributeLines: &attributeLines,
-                                  collectingInto: &result)
-            }
-        }
-    }
-
-    // Collects all the line numbers containing comments or #if/#endif
-    private func skippedLineNumbers(file: SwiftLintFile) -> Set<Int> {
-        var result = Set<Int>()
-        let syntaxMap = file.syntaxMap
-
-        for token in syntaxMap.tokens where token.kind == .comment ||
-                                            token.kind == .docComment {
-            let startLine = file.line(byteOffset: token.offset)
-            var endLine = file.line(byteOffset: token.offset + token.length)
-
-            if file.contents(for: token)?.starts(with: "/*") == true {
-                endLine += 1
-            }
-
-            if startLine <= endLine {
-                result.formUnion(Set(startLine...endLine))
-            }
-        }
-
-        let directiveLines = file.lines.filter {
-            return regex(#"^\s*#(if|elseif|else|endif|\!|warning|error)"#)
-                .firstMatch(in: $0.content, options: [], range: $0.content.fullNSRange) != nil
-        }
-
-        result.formUnion(directiveLines.map { $0.index - 1 })
-        return result
-    }
-
-    // Collects all the line numbers containing attributes but not declarations
-    // other than let/var
-    private func attributeLineNumbers(file: SwiftLintFile) -> Set<Int> {
-        let lineNumbers = file.syntaxMap.tokens
-            .filter { isAttribute(token: $0, in: file) }
-            .map(\.offset)
-            .compactMap(file.line)
-        return Set(lineNumbers)
-    }
-
-    private func isAttribute(token: SwiftLintSyntaxToken, in file: SwiftLintFile) -> Bool {
-        let kind = token.kind
-        if kind == .attributeBuiltin {
-            return true
-        }
-        if kind == .typeidentifier, let symbol = file.stringView.substringStartingLinesWithByteRange(token.range) {
-            return symbol.trimmingCharacters(in: .whitespaces).starts(with: "@")
-        }
-        return false
+    private static func wrapIntoClass(_ example: Example) -> Example {
+        example.with(code: "class C {\n" + example.code + "\n}")
     }
 }
 
-private extension SwiftDeclarationKind {
-    // The various kinds of let/var declarations
-    static let varKinds: [SwiftDeclarationKind] = [.varGlobal, .varClass, .varStatic, .varInstance]
-    // Declarations other than let & var that can have attributes
-    static let nonVarAttributableKinds: [SwiftDeclarationKind] = [
-        .class, .struct,
-        .functionFree, .functionSubscript, .functionDestructor, .functionConstructor,
-        .functionMethodClass, .functionMethodStatic, .functionMethodInstance,
-        .functionOperator, .functionOperatorInfix, .functionOperatorPrefix, .functionOperatorPostfix ]
+private extension LetVarWhitespaceRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visitPost(_ node: MemberBlockItemListSyntax) {
+            if node.parent?.is(IfConfigClauseSyntax.self) != false {
+                return
+            }
+            for member in node {
+                let decl = member.decl
+                guard !decl.is(MacroExpansionDeclSyntax.self),
+                      let index = node.index(of: member),
+                      case let nextIndex = node.index(after: index),
+                      nextIndex != node.endIndex,
+                      case let nextDecl = node[node.index(after: index)].decl,
+                      !nextDecl.is(MacroExpansionDeclSyntax.self) else {
+                    continue
+                }
+                if decl.kind != nextDecl.kind, decl.kind == .variableDecl || nextDecl.kind == .variableDecl,
+                   !(decl.trailingTrivia + nextDecl.leadingTrivia).containsAtLeastTwoNewlines {
+                    violations.append(nextDecl.positionAfterSkippingLeadingTrivia)
+                }
+            }
+        }
+    }
 }
 
-private extension SwiftLintFile {
-    // Zero based line number for specified byte offset
-    func line(byteOffset: ByteCount) -> Int {
-        let lineIndex = lines.firstIndexAssumingSorted { line in
-            return line.byteRange.location > byteOffset
-        }
-        return (lineIndex ?? 0 ) - 1
+private extension Trivia {
+    var containsAtLeastTwoNewlines: Bool {
+        reduce(into: 0) { result, piece in
+            if case let .newlines(number) = piece {
+                result += number
+            }
+        } > 1
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/LetVarWhitespaceRule.swift
@@ -45,6 +45,7 @@ struct LetVarWhitespaceRule: OptInRule {
             Example("""
                 #if os(macOS)
                 let a = 0
+
                 func f() {}
                 #endif
             """),
@@ -120,7 +121,25 @@ struct LetVarWhitespaceRule: OptInRule {
                 @Flag(help: "help")
                 var useAlternativeExcluding = false
             """#, excludeFromDocumentation: true)
-        ].map(Self.wrapIntoClass),
+        ].map(Self.wrapIntoClass) + [
+            Example("""
+                a = 2
+            """),
+            Example("""
+                a = 2
+
+                var b = 3
+            """),
+            Example("""
+                #warning("message")
+                let a = 2
+            """),
+            Example("""
+                #if os(macOS)
+                let a = 2
+                #endif
+            """)
+        ],
         triggeringExamples: [
             Example("""
                 let a
@@ -147,8 +166,19 @@ struct LetVarWhitespaceRule: OptInRule {
                 @Wapper
                 var isEnabled = true
                 ↓func g() {}
+            """),
+            Example("""
+                #if os(macOS)
+                let a = 0
+                ↓func f() {}
+                #endif
             """)
-        ].map(Self.wrapIntoClass)
+        ].map(Self.wrapIntoClass) + [
+            Example("""
+                let a = 2
+                ↓b = 1
+            """)
+        ]
     )
 
     private static func wrapIntoClass(_ example: Example) -> Example {
@@ -159,22 +189,32 @@ struct LetVarWhitespaceRule: OptInRule {
 private extension LetVarWhitespaceRule {
     final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
         override func visitPost(_ node: MemberBlockItemListSyntax) {
-            if node.parent?.is(IfConfigClauseSyntax.self) != false {
-                return
+            collectViolations(from: node, using: { $0.decl })
+        }
+
+        override func visitPost(_ node: CodeBlockItemListSyntax) {
+            if node.parent?.is(SourceFileSyntax.self) == true {
+                collectViolations(from: node, using: { $0.unwrap })
             }
-            for member in node {
-                let decl = member.decl
-                guard !decl.is(MacroExpansionDeclSyntax.self),
-                      let index = node.index(of: member),
-                      case let nextIndex = node.index(after: index),
-                      nextIndex != node.endIndex,
-                      case let nextDecl = node[node.index(after: index)].decl,
-                      !nextDecl.is(MacroExpansionDeclSyntax.self) else {
+        }
+
+        private func collectViolations<List: SyntaxCollection>(from members: List,
+                                                               using unwrap: (List.Element) -> any SyntaxProtocol) {
+            for member in members {
+                guard case let item = unwrap(member),
+                      !item.is(MacroExpansionDeclSyntax.self),
+                      !item.is(MacroExpansionExprSyntax.self),
+                      let index = members.index(of: member),
+                      case let nextIndex = members.index(after: index),
+                      nextIndex != members.endIndex,
+                      case let nextItem = unwrap(members[members.index(after: index)]),
+                      !nextItem.is(MacroExpansionDeclSyntax.self),
+                      !nextItem.is(MacroExpansionExprSyntax.self) else {
                     continue
                 }
-                if decl.kind != nextDecl.kind, decl.kind == .variableDecl || nextDecl.kind == .variableDecl,
-                   !(decl.trailingTrivia + nextDecl.leadingTrivia).containsAtLeastTwoNewlines {
-                    violations.append(nextDecl.positionAfterSkippingLeadingTrivia)
+                if item.kind != nextItem.kind, item.kind == .variableDecl || nextItem.kind == .variableDecl,
+                   !(item.trailingTrivia + nextItem.leadingTrivia).containsAtLeastTwoNewlines {
+                    violations.append(nextItem.positionAfterSkippingLeadingTrivia)
                 }
             }
         }
@@ -188,5 +228,15 @@ private extension Trivia {
                 result += number
             }
         } > 1
+    }
+}
+
+private extension CodeBlockItemSyntax {
+    var unwrap: any SyntaxProtocol {
+        switch item {
+        case let .decl(decl): decl
+        case let .stmt(stmt): stmt
+        case let .expr(expr): expr
+        }
     }
 }

--- a/Source/SwiftLintCore/Extensions/SwiftLintFile+Cache.swift
+++ b/Source/SwiftLintCore/Extensions/SwiftLintFile+Cache.swift
@@ -10,6 +10,7 @@ import SwiftParserDiagnostics
 import SwiftSyntax
 
 private typealias FileCacheKey = UUID
+
 private let responseCache = Cache { file -> [String: any SourceKitRepresentable]? in
     do {
         return try Request.editorOpen(file: file.file).sendIfNotDisabled()

--- a/Source/SwiftLintCore/Models/LinterCache.swift
+++ b/Source/SwiftLintCore/Models/LinterCache.swift
@@ -20,9 +20,9 @@ private struct FileCache: Codable {
 public final class LinterCache {
     private typealias Encoder = PropertyListEncoder
     private typealias Decoder = PropertyListDecoder
-    private static let fileExtension = "plist"
-
     private typealias Cache = [String: FileCache]
+
+    private static let fileExtension = "plist"
 
     private var lazyReadCache: Cache
     private let readCacheLock = NSLock()

--- a/Source/swiftlint/Helpers/LintOrAnalyzeArguments.swift
+++ b/Source/swiftlint/Helpers/LintOrAnalyzeArguments.swift
@@ -25,7 +25,7 @@ struct LintOrAnalyzeArguments: ParsableArguments {
         Should reformat the Swift files using the same mechanism used by Xcode (via SourceKit).
         Only applied with `--fix`/`--autocorrect`.
         """)
-    var format = false // swiftlint:disable:this let_var_whitespace - the multiline @Flag breaks us
+    var format = false
     @Flag(help: "Use an alternative algorithm to exclude paths for `excluded`, which may be faster in some cases.")
     var useAlternativeExcluding = false
     @Flag(help: "Read SCRIPT_INPUT_FILE* environment variables as files.")

--- a/Tests/SwiftLintFrameworkTests/RuleConfigurationDescriptionTests.swift
+++ b/Tests/SwiftLintFrameworkTests/RuleConfigurationDescriptionTests.swift
@@ -2,8 +2,6 @@
 import SwiftLintTestHelpers
 import XCTest
 
-// swiftlint:disable:next blanket_disable_command
-// swiftlint:disable let_var_whitespace
 // swiftlint:disable file_length
 
 // swiftlint:disable:next type_body_length

